### PR TITLE
observe methods from attribute

### PR DESCRIPTION
**Pull request by brettmc**
Add the capability to observe/hook methods and functions by adding a WithSpan attribute to the function or its interface.
Add the capability to pass function/method arguments to the pre callback by adding a SpanAttribute attribute to a function's parameters or its interface.
The attribute is provided by the extension, as `OpenTelemetry\Instumentation\WithSpan` and `OpenTelemetry\Instumentation\SpanAttribute`.

Two extra parameters are passed to pre hooks:
1. an array which can be used to set values on the span: name, span kind
2. an array of function/method arguments to be used as params (from optional SpanAttribute attribute applied to function/method params), and from WithSpan's attribute argument

There is a restriction that WithSpan hooks can only be added to functions/methods that are not already hooked by something else (because the hooks are added at runtime: when a method is executed and has no hooks, we then check for attribute and add hooks).

Lastly, the default pre/post hook methods are \OpenTelemetry\API\Instrumentation\WithSpanHandler::pre and ::post (assuming we would provide a default implementation in our API, which would be a basic version of an auto-instrumentation module). The defaults can be changed via .ini, so folks can roll their own callbacks.

Possible improvements:

- [x] configurable pre/post hooks from ini
- [x] disable feature from ini
- [x] add stubs for new attributes
- [x] disabled by default?
- [x] combine attribute from WithSpan and SpanAttribute into one array?
- [x] support attribute args from interfaces

Issues:
- one minor issue with generating opentelemetry_arginfo.h from stubs. The parser it uses doesn't like attributes, and I also couldn't get it to generate constructors for the attributes. So, I've just implemented as much as (mostly) works, and documented the post-generation hacks.

Closes: https://github.com/open-telemetry/opentelemetry-php/issues/1355

### CODE REVIEW

**Comment by brettmc** [comment]
one more improvement here I thought of today, is that we could also add attributes to control the callback method/s for a function... eg `WithPreSpanCallback` or something like that? I think it provides greater flexibility than just having one set of callbacks for everything...

File: ext/otel_observer.c [hunk 1]
```diff
@@ -192,6 +298,50 @@ static inline void func_get_lineno(zval *zv, zend_execute_data *ex) {
     }
 }
 
+static inline void func_get_attribute_args(zval *zv, HashTable *attributes,
+                                           zend_execute_data *ex) {
+    if (!OTEL_G(attr_hooks_enabled)) {
+        ZVAL_EMPTY_ARRAY(zv);
+        return;
+    }
+    zend_attribute *attr = find_withspan_attribute(ex->func);
```
Comment (agoallikmaa) [inline ext/otel_observer.c:307]: Can you think of any potential issues with saving this during resolve time in `otel_observer` structure to reduce the extra work here? With multiple potential interfaces it might not be a very cheap operation.

File: ext/otel_observer.c [hunk 2]
```diff
@@ -432,16 +582,22 @@ static void observer_begin(zend_execute_data *execute_data, zend_llist *hooks) {
...
+    zval params[8];
+    uint32_t param_count = 8;
+    HashTable *attributes;
+    ALLOC_HASHTABLE(attributes);
+    zend_hash_init(attributes, 0, NULL, ZVAL_PTR_DTOR, 0);
 
     func_get_this_or_called_scope(&params[0], execute_data);
-    func_get_args(&params[1], execute_data);
+    func_get_attribute_args(&params[6], attributes, execute_data);
+    func_get_args(&params[1], attributes, execute_data, true);
```
Comment (agoallikmaa) [inline ext/otel_observer.c:593]: If I'm reading this right, this function seems to collect attributes even if `WithSpan` was not present. Is there any point populating this if `WithSpan` was not used? If not, skipping that would save some time on collecting information that will never be used.

File: ext/otel_observer.c [hunk 3]
```diff
@@ -800,7 +968,39 @@ static otel_observer *resolve_observer(zend_execute_data *execute_data) {
 
     if (!zend_llist_count(&observer_instance.pre_hooks) &&
         !zend_llist_count(&observer_instance.post_hooks)) {
-        return NULL;
+        if (OTEL_G(attr_hooks_enabled) && has_withspan_attribute) {
```
Comment (agoallikmaa) [inline ext/otel_observer.c:971]: Perhaps there should be some log message somewhere to at least let people know that `WithSpan` was ignored? Also maybe we should discuss potentially allowing both in the future, but it's definitely safer to initially go with a solution where this is not possible.

File: ext/tests/span_attribute/function_params.phpt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+#[WithSpan]
+function foo(
+    #[SpanAttribute('renamed_one')] string $one,
+    #[SpanAttribute] int $two,
+    #[SpanAttribute('renamed_three')] float $three,
+    #[SpanAttribute] bool $four,
+    string $five,
+    #[SpanAttribute] string $six
+): void
```
Comment (agoallikmaa) [inline ext/tests/span_attribute/function_params.phpt:33]: Perhaps there should also be a test to see what happens if null, array, object or callable are passed for paramters with `SpanAttribute`.

File: ext/otel_observer.c (hunk 1 above)
Comment (brettmc) [reply ext/otel_observer.c:307]: I thought that since all interfaces were already available in `func->common.scope` that it would be fairly cheap to process them. I don't see any issues with stashing data in `otel_observer` (I assume you mean as a cache of functions to not re-check).

File: ext/otel_observer.c (hunk 2 above)
Comment (brettmc) [reply ext/otel_observer.c:593]: Yes, that's true. I think that 4th param should consider if feature is enabled && WithSpan is used here...

File: ext/otel_observer.c (hunk 3 above)
Comment (brettmc) [reply ext/otel_observer.c:971]: I'll add a log message... not sure what you mean by "allowing both" ?

File: ext/tests/span_attribute/function_params.phpt (hunk 4 above)
Comment (brettmc) [reply ext/tests/span_attribute/function_params.phpt:33]: added a test for extra types

Comment (brettmc) [reply ext/tests/span_attribute/function_params.phpt:33]: added a warning message

File: ext/otel_observer.c (hunk 3 above)
Comment (brettmc) [reply ext/otel_observer.c:979]: added a warning message

File: ext/otel_observer.c (hunk 2 above)
Comment (brettmc) [reply ext/otel_observer.c]: updated to only collect attribute if feature enabled and WithSpan exists on the function (or an interface) - this lookup could be another instance where keeping track of which function have/haven't got WithSpan...

**Review by cedricziel — CHANGES_REQUESTED** [review]
I love how this evolved - cannot wait to start annotating :) 

I think we should reconsider the namespace and rather make it part of [`OpenTelemetry/API`](https://packagist.org/packages/open-telemetry/api) so it can always be available in any project that uses the api package so users dont accidentally run into issues when the add attributes, but the extension is not loaded. Note that would mean we need to maintain BC.

**Comment by brettmc** [comment]
> I think we should reconsider the namespace and rather make it part of [OpenTelemetry/API](https://packagist.org/packages/open-telemetry/api) so it can always be available in any project that uses the api package so users dont accidentally run into issues when the add attributes, but the extension is not loaded. Note that would mean we need to maintain BC.

Do you mean that the attributes themselves should live in OpenTelemetry/API namespace, and be provided by the API rather than by the extension? The behaviour when attributes are added but the extension is not installed is that nothing happens. I do think that the attributes would be easier to find if they were in the API, at least. Otherwise, we're relying on IDE stubs...

**Review by agoallikmaa — APPROVED** [review]
Looks good! There is potential future performance optimization work that could be done either proactively or if performance ever comes up as an issue.

**Comment by cedricziel** [comment]
> Do you mean that the attributes themselves should live in OpenTelemetry/API namespace, and be provided by the API rather than by the extension? The behaviour when attributes are added but the extension is not installed is that nothing happens. I do think that the attributes would be easier to find if they were in the API, at least. Otherwise, we're relying on IDE stubs...

Yes, i think the attributes themselves should live in the API package and should be provided by the API package.

My concern is about the classes that are attempted to be loaded by the autoloader - if the extension is not installed, that will fail. So even if users use `--ignore-platform-reqs` with composer, any framework that has a compile step like Symfony and Laravel will fail, even in dev and test, simply because the attribute classes will not exist if the extension is missing. Hence my proposal to add the Attribute classes to the API package so they are auto-loadable in any case, even if the extension is not loaded.

IMHO this behavior ensures smoother sailing with modern frameworks. - A comparably small side-effect is that there is no effect if the extension is not loaded, but that is the pure nature of attributes in my opinion and i would prefer any time of the day over imposing the existence of the extension in every single build and deploy step where i need to execute some app code :)

**Comment by brettmc** [comment]
> Yes, i think the attributes themselves should live in the API package and should be provided by the API package.

Discussed at yesterday's SIG and agreed: https://github.com/open-telemetry/opentelemetry-php/pull/1371

**Comment by brettmc** [comment]
@agoallikmaa @cedricziel - all green again here. Today I have made the following changes which may be worth re-reviewing:

1. removed attributes (and added to the API, as discussed in SIG and above)
2. fixed an issue with generating the `hook` function from stubs + arginfo
3. tidied tests
4. documentation updates

**Comment by jonnyynnoj** [comment]
This is really nice, I ended up implementing something similar in my project so it's good to see this become part of the core extension.

One suggestion I have (and I don't know if this is feasible), is whether the extension could look for an interface instead of specifically for the WithSpan/SpanAttribute classes? The interface would define pre/post methods where the actual logic for creating a span, adding an attribute or anything else would live. That'd provide greater flexibility allowing users to define their own attributes and move logic back into PHP land where it's a bit more approachable.

So the WithSpan attribute could look something like this:

```
#[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_FUNCTION)]
readonly class WithSpan implements \OpenTelemetry\API\Instrumentation\HookAttribute
{
    public function __construct(public ?string $name = null)
    {
    }

    public function pre(mixed $object, array $params, string $class, string $function): void
    {
       // start span
    }

    public function post(array $params, mixed $return, ?\Throwable $exception): void
    {
        // end span
    }
}
```

**Comment by ChrisLightfootWild** [comment]
> This is really nice, I ended up implementing something similar in my project so it's good to see this become part of the core extension.
> 
> One suggestion I have (and I don't know if this is feasible), is whether the extension could look for an interface instead of specifically for the WithSpan/SpanAttribute classes? The interface would define pre/post methods where the actual logic for creating a span, adding an attribute or anything else would live. That'd provide greater flexibility allowing users to define their own attributes and move logic back into PHP land where it's a bit more approachable.
> 
> So the WithSpan attribute could look something like this:
> 
> ```
> #[Attribute(Attribute::TARGET_METHOD | Attribute::TARGET_FUNCTION)]
> readonly class WithSpan implements \OpenTelemetry\API\Instrumentation\HookAttribute
> {
>     public function __construct(public ?string $name = null)
>     {
>     }
> 
>     public function pre(mixed $object, array $params, string $class, string $function): void
>     {
>        // start span
>     }
> 
>     public function post(array $params, mixed $return, ?\Throwable $exception): void
>     {
>         // end span
>     }
> }
> ```

There's a mechanism for [customising the handlers](https://github.com/opentelemetry-php/ext-opentelemetry/blob/1.1.0/tests/with_span/customize_handlers.phpt) already, but not to say what you proposed above isn't feasible - I'd have to defer that to the greater minds!
